appcypher-syntax.md

```diff
@@ -0,0 +1,947 @@
+
+    This is useful for having heterogenous and sophisticated type inference implementation.
+
+    ```swift
+    let identification: String|Int = "120232435" // `identification` has type `String` here
+    identification = 120232435 // `identification` now has type `Int` here
+    ```
+
+    ```swift
+    func print_id(id: String|Int) {
+        print(id)
+    }
+
+    print(identification)
+    ```
+
+
+## NUMERIC LITERALS
+- Ratio has 13 scalar types
+
+    #### Signed Integers
+    - Int - Pointer-sized signed integer
+    - I8 - 8-bit signed integer
+    - I16 - 16-bit signed integer
+    - I32 - 32-bit signed integer
+    - I64 - 64-bit signed integer
+    - Int - Pointer-sized integer
+
+    #### Unsigned Integers
+    - UInt - Pointer-sized unsigned integer
+    - U8 - 8-bit unsigned integer
+    - U16 - 16-bit unsigned integer
+    - U32 - 32-bit unsigned integer
+    - U64 - 64-bit unsigned integer
+
+    #### Floating point numbers
+    - F32 - 32-bit IEEE 754-2019 floating point representation
+    - F64 - 64-bit IEEE 754-2019 floating point representation
+
+- Some examples of numeric literals.
+
+    ```js
+    var index = 5 // UInt literal
+    var axis = -3 // Int literal
+    let meters = 0.25e-5 // F64 literal
+    ```
+
+- Hex, binary and octal literals are supported as well.
+
+    ```js
+    var price = 0x6FFF00p+12 // Hex literal
+    var op_code = 0b10110001 // Binary literal
+    var interest = 0o56676 // Octal literal
+    ```
+
+## TUPLES
+- Tuple is a special datatype with statically-known length. It allows for cool features like arguments unpacking.
+
+    ```swift
+    var fave_game = ("Uncharted 4", 2016)
+    ```
+
+    One-element tuple. Note the trailing comma.
+
+    ```rust
+    let map = (london,)
+    ```
+
+    Empty tuple.
+
+    ```swift
+    let empty_tuple = ()
+    ```
+
+- Tuple indexing. Tuple index space starts at index 0.
+
+    ```rust
+    let book = ("Harry Potter", "J.K. Rowling", 1995)
+    let title = book.1 // Harry Potter
+    let author = book.2 // J.K. Rowling
+    let year = book.3 // 1995
+    ```
+
+- Tuple destructuring.
+
+    ```rust
+    let person = ("Sam", 50)
+    let (name, age) = person
+    ```
+
+- Tuple unpacking.
+
+    ```swift
+    func add(a, b) {
+        a + b
+    }
+
+    let arguments = (5, 6)
+    add(...arguments)
+    ```
+
+- Tuples have limitations because they are expected to be amenable to static analysis.
+
+    ```swift
+    var tup1 = (1, 2)
+    var tup2 = (5, 6)
+    ```
+
+    Directly or indirectly, a tuple can't be used to extend itself.
+
+    ```go
+    for i in range(x) {
+        tup2 = (...tup2, ...tup1) // Invalid!
+    }
+
+    for i in range(x) {
+        tup3 = (...tup2, 1)
+        tup2 = (tup3, tup1) // Invalid!
+    }
+    ```
+
+- The type of a tuple.
+
+    Since tuples can't change in size, a tuple has a fixed type. For example the tuple `(0xFF, "Hi")` has type `(Int, String)`. `("Hello", 5.0, "Hi")` has type `(String, F64, String)`
```

Comment:
Well `tuple` isn't just an immutable list here. Immutable list is easily implemented in the language.

The reason tuple has to be special here is because it has compile-time properties that makes it useful for certain scenarios. 

1.) It has a statically known length
2.) It has a statically known element type

And those are the reasons we can't just make tuple an immutable list. List length, immutable or not, are not always statically known.

```
let length = input_from_user()
let values = List::new(length)
```

We don't know the list length at compile-time, because `length` is only provided at runtime.

Also, for a heterogenous list, the individual element types are not always statically known.

```
let samples: [String|Int] = []

for i in 0..len {
    samples.push(
         if cond { "Hi" } else { 20 }
    )
}

let values: [String|Int] = samples
```

`values` can contain both String and Int elements, but it is very possible that at the end of the loop, there is no item in `values`, that `values` contain only String items, that `values` contain only Int items; that `values` contain items of both types in any order.

```rust
let samples: [String|Int] = []

for i in 0..len {
    samples.push(
         if cond { "Hi" } else { 20 }
    )
}

let values: [String|Int] = samples

func show(a: Int, b: Str) -> {
    print(a², b.upper())
}

show(...values) # Invalid! 
```

We can't unpack `values` here because `values` can have any length here, and its elements can be of any type. But we need those information at compile-time to correctly type-check the function. This is why lists are unsuitable for this.

On the other hand, a tuple has this information statically available. The individual element types and the length of the tuple are known at compile time. 

But being able to unpack as function arguments isn't the only benefit of tuples. They are more efficient than List memory-wise and perf-wise (reduces cache misses). Tuples are like structs, except indexable by integer. 

This is also how Rust and Swift tuples work. Also this is how C++ tuples work, although behind a variadic template type.